lrauv_ignition_plugins/test/test_elevator.cc

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <chrono>
+#include <gtest/gtest.h>
+
+#include "helper/LrauvTestFixture.hh"
+#include "lrauv_command.pb.h"
+
+#include <fstream>
+
+//////////////////////////////////////////////////
+TEST_F(LrauvTestFixture, Elevator)
+{
+  this->fixture->Server()->Run(true, 100, false);
+  EXPECT_EQ(100, this->iterations);
+  EXPECT_EQ(100, this->tethysPoses.size());
+  EXPECT_NEAR(0.0, this->tethysPoses.back().Pos().X(), 1e-6);
+  EXPECT_NEAR(0.0, this->tethysPoses.back().Pos().Y(), 1e-6);
+  EXPECT_NEAR(0.0, this->tethysPoses.back().Pos().Z(), 2e-2);
+
+  // Propel vehicle
+  lrauv_ignition_plugins::msgs::LRAUVCommand cmdMsg;
+
+  // Move forward
+  cmdMsg.set_propomegaaction_(30);
+
+  // Rotate elevator counter-clockwise when looking from starboard, which
+  // causes the vehicle to pitch down
+  cmdMsg.set_elevatorangleaction_(0.27);
```

Comment:
Given the stall angle of the air foil is 0.17 radians as defined here: https://github.com/osrf/lrauv/blob/1181aa3c4515ed131d0994fd4b443c5017144157/lrauv_description/models/tethys_equipped/model.sdf#L118 I don't think we should expect any reasonable behavior if we move the rudder passed that.

Reply:
Hi @arjo129 @chapulina, 

Here are notes regarding the LRAUV control surfaces along with a plot of their lift/drag coefficients as a function of angle of attack. As you can see, the control surfaces are still generating lift with minimal drag throughout the nominal operation range of the fins (+/- 15 degrees). Drag increases past the nominal range. As Arjo pointed out, alpha stall is found at the edge of the linear range around +/-0.17 rad (or ~9.7 deg). 

![lrauv_lift_drag](https://user-images.githubusercontent.com/7927479/140449717-0e08c41e-900f-4919-bc9b-a4ef52c780d0.png)

**LRAUV control surfaces:**
area: Surface area of the link.

122 cm^2 for one fin.

a0: The initial "alpha" or initial angle of attack. a0 is also the y-intercept of the alpha-lift coefficient curve.

0.0 degrees

cla: The ratio of the coefficient of lift and alpha slope before stall. Slope of the first portion of the alpha-lift coefficient curve.

C_{L\alpha} = 4.13  (alpha in radians)

cda: The ratio of the coefficient of drag and alpha slope before stall.

The coefficient of drag at alpha =0 is 0.03.  The ratio drag/aoa is given below.  Please see Coeffs.png.

C_{D\alpha} = .04/.2 = .2 (alpha in radians)

cp: Center of pressure. The forces due to lift and drag will be applied here.

(3.7, 7.9) cm as shown in the attached diagram OriginalFin.png.  This moves as a function of aoa, as show by the attached diagram cp.png.  Note that cp.png is referenced to mean chord length, not the root chord length.

forward: 3-vector representing the forward direction of motion in the link frame.

[1 0 0]

upward: 3-vector representing the direction of lift or drag.

[0 0 -1]

alpha_stall: Angle of attack at stall point; the peak angle of attack.

9.5 degrees (.17 rad)

cla_stall: The ratio of coefficient of lift and alpha slope after stall. Slope of the second portion of the alpha-lift coefficient curve.

-1.1 (alpha in radians)

cda_stall: The ratio of coefficient of drag and alpha slope after stall.

.030 (alpha in radians)

Reply:
Ohh good point. I see on the controller's config that the rudder angle limit is 15 deg (0.261799 rad), which matches the joint limits. The only stall angle that I found on MBARI's code was the tailcone's stall angle, which is 30 deg. Do you remember where the elevator and rudder stall angles that we're using on lift-drag were taken from? We should document it.

If angles beyond stall result in unreasonable behaviour (at least with our lift-drag plugin), maybe we should limit the joint there?

Reply:
From the above plot. Ben sent it to us on slack.

Reply:
I believe the values for the coeffs in the plugin came from the notes above. These notes are from Rob McEwen, our controls engineer. The lrauv-application already limits the angle of the fins to +/- 15 deg, and that's the range we operate with in the field. I guess it's worth noting that the plot above is for the angle of attack and not the fin angle. Still, plenty of lift should be generated past the alpha stall — I think that's also shown in this Gazebo tutorial: http://gazebosim.org/tutorials?tut=aerodynamics&cat=physics

It might be worth revisiting the coeffs values since we there might be a gap between Rob's understanding and what we're doing with these values in the plugin.

Reply:
@braanan posted 2 mins before me :smile: 

It would be nice to double check all these values just to be sure. For example, not sure if by a "fin" we're considering both sides of the elevator or each of them separately. It would also be nice to double check coordinate frames.

> I don't think we should expect any reasonable behavior if we move the rudder/elevator passed that.

@arjo129 , do you expect that because of some limitation on our `LiftDrag` plugin or just in general?

---

As a side note, I had tried some different `cp` values earlier (we currently have `0 0 0` and the simulation crashed. `cp` above is non-zero, so maybe this is one thing for us to look at.

Reply:
Ok, the behaviour looks more reasonable to me at

### elevator = 0.15 rad, prop omega = 30.0 rad/s

![0 15rad30rads](https://user-images.githubusercontent.com/5751272/140579160-d0e50c70-2fe8-4e1a-8aec-a8e0cbe293e9.gif)

But at lower prop omegas, the vehicle still takes off, is that expected?

Reply:
I moved @braanan 's comment to this wiki page: https://github.com/osrf/lrauv/wiki/Lift-drag

Also updated the test to use 0.15 rad.

Reply:
Thanks for capturing those notes. Interesting — that looks much better! Guess we'll have to tweak the lift-drag plugin to support the full actuator range — as I mentioned, the vehicles run with the full +/- 15 deg range.

Reply:
So I guess the elevators were simply stalling in the previous example?

Reply:
30 rad/s is pretty close to our nominal 1 m/s speed but we typically have good control authority at much lower speeds as well — the vehicle's can easily run at 0.5 m/s and even slightly lower than that

Reply:
Can you get the speed through the water?

Reply:
> Can you get the speed through the water?

* elevator = 0.15 rad, prop omega = 30.0 rad/s
* Plotting XYZ linear velocity in the world frame. It eventually settles around:
    * -0.93 m/s on X
    * -0.01 m/s on Y
    * -0.29 m/s on Z

![velocity](https://user-images.githubusercontent.com/5751272/140588203-c651124a-40a5-4e5a-87d9-fb56ceba4acb.gif)

Reply:
Thanks! Hmm.. are these in the vehicle frame? X velocity magnitude is about right but I'd expect the sign to be positive.

Reply:
Ah, oops — I see there in the world frame!

Reply:
Nice!

Reply:
So liftdrag plugin, should theoretically be giving the correct force values up to +/- 15 degrees. The liftdrag doesn't support any Cp other than Cp (0 0 0).

Reply:
I captured the gist of this conversation in https://github.com/osrf/lrauv/issues/78